### PR TITLE
feat(skate/test-groups): send test group information to frontend

### PR DESCRIPTION
**Pull request by firestack**
This implements access to a user's test groups via storing the information in the database schema, rendering it to the frontend, and providing a utility function to access the data.

---

Closes: https://app.asana.com/0/1200180014510248/1202519045249749/f

### CODE REVIEW

File: test/skate_web/controllers/page_controller_test.exs
Comment (firestack) [inline test/skate_web/controllers/page_controller_test.exs:73]: Does this read okay? would it read more "idomatically" if instead it was like this?
```suggestion
      json =
      	user_struct.test_groups
      	|> Enum.map(& &1.name)
        |> Jason.encode!()
        |> Phoenix.HTML.html_escape()
        |> Phoenix.HTML.safe_to_string()
```

I like it being within the `encode` parentheses because it makes it clear what's happening to the `test_groups` first but maybe there's another way to format it and reduce cognitive overhead via indentation/shape?

File: test/skate_web/controllers/page_controller_test.exs
Comment (firestack) [inline test/skate_web/controllers/page_controller_test.exs:71]: I'm also kinda concerned the test needs to handle the case of having multiple groups or groups returned in different orders?

If so maybe it needs to be a different data type?

File: test/skate_web/controllers/page_controller_test.exs
Comment (firestack) [inline test/skate_web/controllers/page_controller_test.exs:62]: The test succeeds with multple test groups defined, but it's somewhat verbose. I'm wondering if anyone else thinks we should be testing with multiple groups in this test or create another one that does that?

File: test/skate_web/controllers/page_controller_test.exs
Comment (KaylaBrady) [reply test/skate_web/controllers/page_controller_test.exs:73]: I personally tend to find it easier to read pipe chains when they start with a single value, like your suggested change does. The first thing that is happening to the test groups is that they are being mapped to only their name, so for me personally it is easier to see that as a distinct step. 

Much of my elixir preferences were informed by working on a project with "controversial" credo checks enabled, for example, the credo check for [pipe_chain_start](https://github.com/rrrene/credo/blob/master/lib/credo/check/refactor/pipe_chain_start.ex) which applies in this scenario. So I acknowledge that this may be a "controversial" preference, and that particular credo check is not enabled in this repo, so I wouldn't block a PR over it. Over all though, I found credo checks to be a helpful guide for picking up conventions as I was starting to learn elixir, and have adopted or rejected some of the "controversial" ones over time.

File: assets/src/userTestGroups.ts
Comment (lemald) [reply assets/src/userTestGroups.ts:3]: For what it's worth we usually declare TypeScript functions with `const` rather than `function` in most of the codebase.

Comment (lemald) [reply assets/src/userTestGroups.ts:3]: As for exporting I'd say it's fine to leave private for now and we can export it later if need be.

File: assets/src/userTestGroups.ts
Comment (lemald) [inline assets/src/userTestGroups.ts:5]: Since `JSON.parse` returns an `any` type TypeScript doesn't have a problem with this, but it would still be nice to do some runtime checks to make sure the result of parsing is actually an array of strings like we'd expect.

File: test/skate_web/controllers/page_controller_test.exs
Comment (lemald) [reply test/skate_web/controllers/page_controller_test.exs:73]: I also prefer starting the pipe with a single value.

File: assets/src/userTestGroups.ts
Comment (firestack) [reply assets/src/userTestGroups.ts:5]: Typescript can't infer based on the `string[]` annotation?

Could/should this be done with superstruct in some form?

File: assets/src/userTestGroups.ts
Comment (firestack) [reply assets/src/userTestGroups.ts:3]: > For what it's worth we usually declare TypeScript functions with `const` rather than `function` in most of the codebase.

I'm just curious, can you elaborate on why?

I understand that with `const` we get errors on reassignment, but we loose function hoisting, which doesn't matter in this case (and due to modules), but it seems strange to me that const functions would be preferred over the language keyword, but I feel like I'm missing something.

File: test/skate_web/controllers/page_controller_test.exs
Comment (lemald) [reply test/skate_web/controllers/page_controller_test.exs:62]: I'm not overly concerned but testing with 0, 1, and N would certainly be nice. I don't think ExUnit really has built-in test parametrization tools, but your test can always wrap some assertions in a `for` block.

File: assets/src/userTestGroups.ts
Comment (firestack) [reply assets/src/userTestGroups.ts]: I'm curious what you think of the revision!

Comment (lemald) [reply assets/src/userTestGroups.ts]: I hadn't really be familiar with `create` as opposed to `assert` and `is` previously but I think that looks good!

**Comment by lemald** [comment]
@firestack the one thing that sticks out to me that still needs doing is tests for `userTestGroups.ts`, but other than that I think this is close to ready for approval!

**Comment by firestack** [comment]
Writing the typescript test and then I'll have y'all hopefuly able to final review 👍🏻

**Comment by firestack** [comment]
I would still like to make [the elixir test more robust](https://github.com/mbta/skate/pull/1788#discussion_r1014168942) but this PR is ready for review and could be squash-merged as is.

File: assets/tests/userTestGroups.test.ts
Comment (lemald) [inline assets/tests/userTestGroups.test.ts:7]: **nitpick:** looks like this was left in unintentionally.

File: assets/tests/userTestGroups.test.ts
Comment (lemald) [inline assets/tests/userTestGroups.test.ts:23]: **suggestion:** it would also be nice to have a test for the case where the JSON data is malformed (as in it parses as valid JSON but isn't an array of strings).

File: lib/skate/settings/db/user.ex
Comment (lemald) [reply lib/skate/settings/db/user.ex:24]: I don't personally feel like this needs a comment, given that you can see it's used in the `has_many` declaration for `:test_groups`, though I could be convinced otherwise if it still feels highly unintuitive to others.

**Comment by lemald** [comment]
@firestack okay left a couple of minor comments that I'd like to see addressed before merging but it's getting there. 👍

File: lib/skate/settings/db/user.ex
Comment (firestack) [reply lib/skate/settings/db/user.ex:24]: I found myself very confused when first reading over this, but this feels like a special case to elixir beginners.

It's personally confusing how something inside a function call can become a referencable member, but we don't document this in all the other similar places, so I don't believe it needs to stay.

File: assets/tests/userTestGroups.test.ts
Comment (lemald) [inline assets/tests/userTestGroups.test.ts:29]: Sorry, what I meant to get at with my earlier comment wasn't only cases where the value is something that doesn't parse as JSON, but also when it's something like `"{"key": "value"}"` that is valid JSON, but not an array of strings when parsed.

**Review by KaylaBrady — APPROVED** [review]

**Review by lemald — APPROVED** [review]

**Pull request merged**
